juju_verify/verifiers/__init__.py

```diff
@@ -73,5 +73,6 @@ def get_verifiers(units: List[Unit]) -> Iterator[BaseVerifier]:
             )
             continue
 
+        exclude_affected_units = list(set(units).difference(charm_units))
```

Comment:
> I've found that set doesn't work for units at all, and the same thing, if you try

`set()` works based on `__hash__`, not `__eq__`. This test shows that the set works for this scenario (`units` is `List[Unit]`)
```python
print("set_test: %s" % set([units[0], units[0]]))
```
outputs
```
set_test: {<Unit entity_id="ceph-mon/0">}
```

Now, I can't figure out why your test/example does not work. I tested this again on machine that runs 3 units

- ceph-mon/0
- ceph-osd/0
- nova-compute/2

I added this debug output after line 76 (this inline comment)
```python
        print("all units: %s" % units)
        print("excluded: %s" % exclude_affected_units)
```
and outputs are exactly what I expected.

For `ceph-mon` verifier:
```
all units: [<Unit entity_id="ceph-mon/0">, <Unit entity_id="ceph-osd/0">, <Unit entity_id="nova-compute/2">]
excluded: [<Unit entity_id="nova-compute/2">, <Unit entity_id="ceph-osd/0">]
```

For `ceph-osd` verifier:
```
all units: [<Unit entity_id="ceph-mon/0">, <Unit entity_id="ceph-osd/0">, <Unit entity_id="nova-compute/2">]
excluded: [<Unit entity_id="nova-compute/2">, <Unit entity_id="ceph-mon/0">]
```

For `nova-compute` verifier:
```
all units: [<Unit entity_id="ceph-mon/0">, <Unit entity_id="ceph-osd/0">, <Unit entity_id="nova-compute/2">]
excluded: [<Unit entity_id="ceph-osd/0">, <Unit entity_id="ceph-mon/0">]
```

Reply:
I tried it again and problem is only if you are getting units from model again.
```python
exclude_affected_units = list(set(units).difference(charm_units))
logger.info("all units: %s" % units)
logger.info("charm units: %s" % charm_units)
logger.info("excluded: %s" % exclude_affected_units)

model = charm_units[0].model
test_charm_units = [model.units[unit.entity_id] for unit in charm_units]
exclude_affected_units = list(set(units).difference(test_charm_units))
logger.info("all units: %s" % units)
logger.info("charm units: %s" % test_charm_units)
logger.info("excluded: %s" % exclude_affected_units)
```
will end up 
```bash
all units: [<Unit entity_id="ceph-osd/0">, <Unit entity_id="ceph-mon/0">]
charm units: [<Unit entity_id="ceph-mon/0">]
excluded: [<Unit entity_id="ceph-osd/0">]
all units: [<Unit entity_id="ceph-osd/0">, <Unit entity_id="ceph-mon/0">]
charm units: [<Unit entity_id="ceph-mon/0">]
excluded: [<Unit entity_id="ceph-osd/0">, <Unit entity_id="ceph-mon/0">]
```

So the way you used it, it works, but we have to be careful.

For example, I switch from set to dictionary [here](https://github.com/canonical/juju-verify/commit/82545d6e9972b7706cf200f7327171cd1bb2b03d#diff-213e10020062a54e67094b089576569e4f48d2ad9626cbda594f574126a4dab6R453-R455), because caused that the `get-health` action was run twice for `juju-verify reboot --units ceph-osd-ssd/0 ceph-osd-hdd/0` instead of once.